torch/csrc/autograd/variable.h

```diff
@@ -291,6 +292,22 @@ struct TORCH_API Variable : public at::Tensor {
   const std::vector<std::shared_ptr<FunctionPreHook>>& hooks() const noexcept;
   void clear_hooks();
 
+  template <typename T>
+  using hook_return_void_t = c10::guts::enable_if_t<std::is_void<decltype(T(std::declval<Variable>()))>::value, unsigned>;
+  template <typename T>
+  using hook_return_var_t = c10::guts::enable_if_t<std::is_same<decltype(T(std::declval<Variable>())), Variable>::value, unsigned>;
+  // Remove hook at given position
+  void remove_hook(unsigned pos);
+
+  // Returns the index of the hook in the list which can be used to remove hook
+  // Register a hook with no return value
+  template <typename T>
+  hook_return_void_t<T> register_hook(T&& hook);
+  // Register a hook with variable return value
+  template <typename T>
+  hook_return_var_t<T> register_hook(T&& hook);
```

Comment:
If my hook returns an undefined tensor, what happens?

Reply:
It looks like it errors. But maybe you should treat that as "no change requested"? (In analogy to returning `None` from a normal hook.)